compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+                ty::Float(FloatTy::F32) => "u3f32",
+                ty::Float(FloatTy::F64) => "u3f64",
+                _ => "",
+            });
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Char => {
+            // u4char as vendor extended type
+            let mut s = String::from("u4char");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Str => {
+            // u3str as vendor extended type
+            let mut s = String::from("u3str");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Never => {
+            // u5never as vendor extended type
+            let mut s = String::from("u5never");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Compound types
+        // () in Rust is equivalent to void return type in C
+        ty::Tuple(_) if ty.is_unit() => {
+            typeid.push('v');
+        }
+
+        // Sequence types
+        ty::Tuple(tys) => {
+            // u5tupleI<element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u5tupleI");
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, options));
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // A<array-length><element-type>
+            let mut s = String::from("A");
+            s.push_str(&format!("{}", &len.val().try_to_scalar().unwrap().to_i64().unwrap()));
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        ty::Adt(adt_def, substs) => {
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.item_name(def_id));
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+            // respective builtin type encoding as specified by the Itanium C++ ABI if
+            // encoding C types is enabled.
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                // C char, integer, and void types
+                s.push('v');
```

Comment:
This should be a separate transformation step, independent of mangling.

Right now in the example below, `f1` and `g1` have the same typeid, while for `f2` and `g2` typeids are distinct.

```rust
pub fn f1(_: *const ()) {}                // _ZTSFvPKvE
pub fn g1(_: *const core::ffi::c_void) {} // _ZTSFvPKvE

pub fn f2(_: *const (), _: *const ()) {}                // _ZTSFvPKvS0_E
pub fn g2(_: *const (), _: *const core::ffi::c_void) {} // _ZTSFvPKvPKvE
```

Reply:
Thanks for pointing it out! I'll fix it.

Reply:
Done.